### ISSUE TITLE
BufferManager must be installed before read interest is registered

### ISSUE THREAD

**Issue by myronkscott** [Terracotta-OSS/terracotta-core#772]
When bootstrapping a connection, the buffer manager for the connection must be installed before the read interest is registered.

**Comment by myronkscott** [comment]
Problem was introduced here.

https://github.com/Terracotta-OSS/terracotta-core/commit/99bbd8dc6eaffc2b7b83f970f44efec2bae947d4

**Comment by ljacomet** [comment]
Requesting that this be fixed both in 5.3 and 5.4

**Comment by myronkscott** [comment]
While the code review shows that the arrangement of calls is indeed incorrect, the possibility of this causing a problem in practice is near zero.  When the connection is opened, the handshake is initiated only after the connection is fully established.  This race could only occur if there are unexpected packets being presented to the ephemeral port just opened.  In such instances, what is being read on the port will cause undefined behavior.  This will be fixed but should be treated with low priority.

**Comment by ewirch** [comment]
I must be very unlucky. This hits me 100% of the time with this configuration and Ehcache 3.4.0:

```
<ehcache:config
		xmlns:xsi='http://www.w3.org/2001/XMLSchema-instance'
		xmlns:ehcache='http://www.ehcache.org/v3'
		xmlns:terracotta='http://www.ehcache.org/v3/clustered'>

	<ehcache:service>
		<terracotta:cluster>
			<terracotta:connection url="terracotta://localhost:9510/clustered"/>
			<terracotta:server-side-config auto-create="true">
				<terracotta:default-resource from="default-resource"/>
			</terracotta:server-side-config>
		</terracotta:cluster>
	</ehcache:service>

	<ehcache:cache alias="myCache">
		<ehcache:key-type>java.lang.String</ehcache:key-type>
		<ehcache:value-type>java.lang.String</ehcache:value-type>

		<ehcache:resources>
			<ehcache:heap unit="entries">200</ehcache:heap>
			<terracotta:clustered-dedicated unit="MB">5</terracotta:clustered-dedicated>
		</ehcache:resources>
	</ehcache:cache>
</ehcache:config>
```

Ehcache always fails to start with a `NullPointerException`.

**Comment by ljacomet** [comment]
@ewirch So what you are saying is that when seeing that exception (could you post the full stack trace btw?), the Ehcache client application dies?

**Comment by myronkscott** [comment]
@ewirch I suspect that your client may not be opening a connection to TSA or the correct version of TSA.  TSA default port changed to 9410 recently.  Can you verify that TSA is listening on port 9510?  If you could attach the server log and the tc-config.xml that you are using, it might help figure out what is going on.

**Comment by ewirch** [comment]
@ljacomet No the application does not die. The `NullPointerException` is being catched by the `L1_L2:TCComm Main Selector Thread_R`. Here are the stack traces I originally posted on [Google Groups](https://groups.google.com/d/msg/ehcache-users/WXIyVEiKMww/bDaJi2RqDAAJ):


`com.tc.net.core.TCConnectionImpl#bufferManager` is read and written in separate threads which run concurrently at the same time. This stack trace starts the initialization thread:

```
at com.tc.object.DistributedObjectClient.start(DistributedObjectClient.java:249)
at com.tc.object.DistributedObjectClientFactory.create(DistributedObjectClientFactory.java:103)
at com.terracotta.connection.TerracottaInternalClientImpl.init(TerracottaInternalClientImpl.java:76)
at com.terracotta.connection.api.TerracottaConnectionService.connect(TerracottaConnectionService.java:99)
at org.terracotta.connection.ConnectionFactory.connect(ConnectionFactory.java:43)
at org.ehcache.clustered.client.internal.service.DefaultClusteringService.initClusterConnection(DefaultClusteringService.java:194)
at org.ehcache.clustered.client.internal.service.DefaultClusteringService.start(DefaultClusteringService.java:144)
at org.ehcache.core.internal.service.ServiceLocator.startAllServices(ServiceLocator.java:118)
at org.ehcache.core.EhcacheManager.init(EhcacheManager.java:560)
```

This stack trace starts the thread which uses `bufferManager`:

```
at com.tc.net.core.CoreNIOServices.<init>(CoreNIOServices.java:88)
at com.tc.net.core.TCCommImpl.<init>(TCCommImpl.java:48)
at com.tc.net.core.TCConnectionManagerImpl.<init>(TCConnectionManagerImpl.java:79)
at com.tc.net.protocol.tcm.CommunicationsManagerImpl.<init>(CommunicationsManagerImpl.java:207)
at com.tc.object.StandardClientBuilder.createCommunicationsManager(StandardClientBuilder.java:72)
at com.tc.object.DistributedObjectClient.start(DistributedObjectClient.java:282)
at com.tc.object.DistributedObjectClientFactory.create(DistributedObjectClientFactory.java:103)
at com.terracotta.connection.TerracottaInternalClientImpl.init(TerracottaInternalClientImpl.java:76)
at com.terracotta.connection.api.TerracottaConnectionService.connect(TerracottaConnectionService.java:99)
at org.terracotta.connection.ConnectionFactory.connect(ConnectionFactory.java:43)
at org.ehcache.clustered.client.internal.service.DefaultClusteringService.initClusterConnection(DefaultClusteringService.java:194)
at org.ehcache.clustered.client.internal.service.DefaultClusteringService.start(DefaultClusteringService.java:144)
at org.ehcache.core.internal.service.ServiceLocator.startAllServices(ServiceLocator.java:118)
at org.ehcache.core.EhcacheManager.init(EhcacheManager.java:560)
```

If the READER thread is faster than the INIT thread, then it will fail with a `NullPointerException`. Which happens for me.

Reader thread reaches this point: 

```
at com.tc.net.core.TCConnectionImpl.doReadInternal(TCConnectionImpl.java:297)
at com.tc.net.core.TCConnectionImpl.doRead(TCConnectionImpl.java:291)
at com.tc.net.core.CoreNIOServices$CommThread.selectLoop(CoreNIOServices.java:674)
at com.tc.net.core.CoreNIOServices$CommThread.run(CoreNIOServices.java:344)
```

faster than the INIT thread reaches this point:

```
at com.tc.net.core.TCConnectionImpl.installBufferManager(TCConnectionImpl.java:236)
at com.tc.net.core.TCConnectionImpl.finishConnect(TCConnectionImpl.java:200)
at com.tc.net.core.TCConnectionImpl.connect(TCConnectionImpl.java:726)
at com.tc.net.protocol.transport.ClientMessageTransport.connect(ClientMessageTransport.java:151)
at com.tc.net.protocol.transport.ClientMessageTransport.open(ClientMessageTransport.java:115)
at com.tc.net.protocol.transport.ClientConnectionEstablisher.connectTryAllOnce(ClientConnectionEstablisher.java:146)
at com.tc.net.protocol.transport.ClientConnectionEstablisher.open(ClientConnectionEstablisher.java:131)
at com.tc.net.protocol.ClientNetworkStackHarness$1.openMessageTransport(ClientNetworkStackHarness.java:98)
at com.tc.net.protocol.tcm.ClientMessageChannelImpl.open(ClientMessageChannelImpl.java:94)
at com.tc.object.DistributedObjectClient.openChannel(DistributedObjectClient.java:448)
at com.tc.object.DistributedObjectClient.access$400(DistributedObjectClient.java:131)
at com.tc.object.DistributedObjectClient$5.run(DistributedObjectClient.java:377)
at java.lang.Thread.run(Thread.java:748)

```

> I suspect that your client may not be opening a connection to TSA or the correct version of TSA. TSA default port changed to 9410 recently. 

@myronkscott That's right the connection fails to establish. I just reproduced it again. The hint with the old port is the important information to trigger the behavior. It occurs when there is some application listening at the port, but it's not Terracotta.

I created and started a Terracotta Docker image as described in your `docker` repo: https://github.com/Terracotta-OSS/docker/tree/master/5.3.2/server The documentation is outdated and uses port 9510:

```
docker run --name tc-server -p 9510:9510 -d terracotta/terracotta-server-oss:5.3.2
```

But Terracotta inside the Docker image is configured to listen to 9410. So the Docker VM started and opened the local port 9510. But no application actually listened on that port inside the Docker VM. When the Ehcache client tries to establish a connection to 9510, the data race and the `NullPointerException` occurs. When I change the setting to an arbitrary unbound port, then a proper `TimeoutException` is being thrown.

**Comment by ewirch** [comment]
> The documentation is outdated and uses port 9510:

Apologies, the documentation is not outdated. But it fails to mention, that one has to define environment variables `TC_SERVER1` and `TC_SERVER2` before starting the Docker container in order to have a Terracotta instance listen on 9510.

**Comment by ljacomet** [comment]
@ewirch Thanks for the detailed explanation.
@anthonydahanne Can you have a look at that docker mismatch?

**Comment by anthonydahanne** [comment]
@ewirch @ljacomet you're right there was some confusion with port numbers in the Dockerfiles, READMEs configs and compose files.
So, now it should be fixed : 9410 and 9430 are consistently the default ports used in terracotta docker images 5.3 and up.
